mono/mini/driver.c

```diff
@@ -2972,40 +3014,63 @@ mono_parse_options (const char *options, int *ref_argc, char **ref_argv [], gboo
 	}
 	if (in_quotes) 
 		return g_strdup_printf ("Unmatched quotes in value: [%s]\n", options);
-		
+
 	if (buffer->len != 0)
 		g_ptr_array_add (array, g_strdup (buffer->str));
 	g_string_free (buffer, TRUE);
 
-	if (array->len > 0){
-		int new_argc = array->len + argc;
-		char **new_argv = g_new (char *, new_argc + 1);
-		int j;
+	merge_parsed_options (array, ref_argc, ref_argv, prepend);
+	g_ptr_array_free (array, TRUE);
 
-		new_argv [0] = argv [0];
+	return NULL;
+}
 
-		i = 1;
-		if (prepend){
-			/* First the environment variable settings, to allow the command line options to override */
-			for (i = 0; i < array->len; i++)
-				new_argv [i+1] = (char *)g_ptr_array_index (array, i);
-			i++;
-		}
-		for (j = 1; j < argc; j++)
-			new_argv [i++] = argv [j];
-		if (!prepend){
-			for (j = 0; j < array->len; j++)
-				new_argv [i++] = (char *)g_ptr_array_index (array, j);
+#if defined(HOST_WIN32) && G_HAVE_API_SUPPORT(HAVE_CLASSIC_WINAPI_SUPPORT)
+#include <shellapi.h>
+
+static char *
+mono_win32_parse_options (const char *options, int *ref_argc, char **ref_argv [], gboolean prepend)
+{
+	int argc;
+	gunichar2 **argv;
+	gunichar2 *optionsw;
+
+	if (!options)
+		return NULL;
+
+	GPtrArray *array = g_ptr_array_new ();
+	optionsw = g_utf8_to_utf16 (options, -1, NULL, NULL, NULL);
+	if (optionsw) {
+		argv = CommandLineToArgvW (optionsw, &argc);
```

Comment:
Likely CommandLineToArgvW is different than the C runtime.
Maybe try to reuse C runtime logic instead?
Dependency on shell32.dll etc. is distasteful too imho.

Reply:
This will mimic the exact same behavior as you will get when passing the arguments to main method on Windows and that's was the whole point. The dependency already exists from other sources, so this is not a new introduced dependency.

Reply:
This does not mimic main/wmain. That was my point. It is different regarding quotes. I'll have to write small samples.

Reply:
What I mean was that it mimic how mono handles arguments on Windows since it doesn't use arguments passed to main/wmain.

Reply:
Oh right, that. Yes that is dubious too. Quotes and spaces are a mess no matter what.
So, there another angle here to consider.
Once you are in a response file, the argv delimiters can be considered newlines and carriage returns instead of spaces. Spaces can be left therefore unquoted.
The "new" difficult to pass characters can become newlines/carriage returns, which are more rare than horizontal spaces.
I'm not sure if there is precedent for that or not. Certainly accepting newlines in response files is normal, the compiler/linker do, and build systems use it. I'm not sure that affects space/quote rules though.

Reply:
True, I will make a try to handle newlines in the file, probably use the parsing logic already available to handle this inside/outside quotes and then let the CommadLineToArgvW take care of the rest.

Reply:
Putting newline in the response file also exposed another issue on Windows and the way g_file_get_contents is implemented, need to adjust that one as well.

Reply:
I did a quick experiment with cl.exe.
My hopes that newlines act as quotes were dashed.
You still need quotes around spaces. So it still a mess.
My assumption that newlines are skipped, or treated as whitespace separating parameters, were preserved.

I don't know how common it is, but some build systems definitely put newlines in compiler/linker response files. It makes the response files somewhat more readable, for those few of us that debug this sort of thing. Perhaps nobody will do that to mono.

Eventually or presumably already, VS and VS4Mac are consumers of this.
So what they want/do is important.

Reply:
Latest commit seems to take care of most scenarios for newline in response file. Preserved when inside quotes, but replaced with whitespace when outside so that shell argument parsing will handle them correctly. Tried with various response file (one was 21 KB) that mixed new lines, white spaces etc and seemed to get expected results. For example (uses new line, tabs, space both inside outside quotes):

```
-argument1 "test
 some more"
-argument2 test2	  	  

 
-argument3 "some	more whitespaces	"
     

Execütable\Program.exe
```

Gets parsed to this (debugger output):

argv[0] L"-argument1"
argv[1] L"test\n some more"
argv[2] L"-argument2"
argv[3] L"test2"
argv[4] L"-argument3"
argv[5] L"some\tmore whitespaces\t"
argv[6] L"Execütable\\Program.exe"